File: sdk/node/examples/example_client.js
#!/usr/bin/env node
// example_client.js — connect to the example server and use every capability.
import path from "node:path";
import { fileURLToPath } from "node:url";
import * as rcp from "../src/index.js";

const here = path.dirname(fileURLToPath(import.meta.url));

async function main() {
  const server = path.join(here, "example_server.js");
  const c = await rcp.connectStdio(["node", server]);

  console.log(`connected to ${c.server.name} v${c.server.version} (RCP/${c.protocolVersion})`);
  console.log("capabilities:", Object.keys(c.capabilities));

  if (c.supports(rcp.Capability.Embed)) {
    const vecs = await c.embed(["hello world"]);
    console.log(`embed      -> ${vecs.length} vector(s), dim=${vecs[0].length}`);
  }

  if (c.supports(rcp.Capability.Retrieve)) {
    console.log("retrieve   ->");
    for (const h of await c.retrieve("landmark in the French capital", 2)) {
      console.log(`   ${h.id}  score=${h.score.toFixed(3)}  ${h.text.slice(0, 48)}`);
    }
  }

  if (c.supports(rcp.Capability.Graph)) {
    const g = await c.graph("global", { query: "anything" });
    console.log("graph      ->", g);
  }

  await c.shutdown();
}

main();

File: sdk/rust/examples/example_client.rs
//! example_client.rs — connect to the example server and use every capability.
//!
//!   cargo build --examples
//!   cargo run --example example_client

use rcp::Capability;

fn main() -> Result<(), rcp::RcpError> {
    // The example_server binary sits next to this one in target/.../examples/.
    let exe = std::env::current_exe().expect("current exe");
    let server = exe.parent().unwrap().join("example_server");
    let server = server.to_string_lossy().to_string();

    let mut c = rcp::connect_stdio(&[server.as_str()])?;

    println!(
        "connected to {} v{} (RCP/{})",
        c.server().get_str("name").unwrap_or("?"),
        c.server().get_str("version").unwrap_or("?"),
        c.protocol_version()
    );
    if let Some(caps) = c.capabilities().as_object() {
        let keys: Vec<&str> = caps.iter().map(|(k, _)| k.as_str()).collect();
        println!("capabilities: {:?}", keys);
    }

    if c.supports(Capability::Embed) {
        let vecs = c.embed(&["hello world"], None)?;
        println!("embed      -> {} vector(s), dim={}", vecs.len(), vecs[0].len());
    }

    if c.supports(Capability::Retrieve) {
        println!("retrieve   ->");
        for h in c.retrieve("landmark in the French capital", 2)? {
            let snippet: String = h.text.chars().take(48).collect();
            println!("   {}  score={:.3}  {}", h.id, h.score, snippet);
        }
    }

    if c.supports(Capability::Graph) {
        let g = c.graph("global", Some(rcp::obj(&[("query", "anything".into())])))?;
        println!("graph      -> {}", g);
    }

    c.shutdown();
    Ok(())
}
